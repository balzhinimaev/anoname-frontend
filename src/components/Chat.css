/* CSS переменные для чата */
:root {
  --chat-bg-light: #ffffff;
  --chat-bg-dark: #2c2c2e;
  --messages-bg-light: #f8f9fc;
  --messages-bg-dark: #1c1c1e;
  --message-my-bg: #007aff;
  --message-partner-bg-light: #f2f2f7;
  --message-partner-bg-dark: #3a3a3c;
  --input-bg-light: #f2f2f7;
  --input-bg-dark: #3a3a3c;
  --input-area-bg-light: #ffffff;
  --input-area-bg-dark: #2c2c2e;
  --text-primary-light: #1d1d1f;
  --text-primary-dark: #f2f2f7;
  --text-secondary-light: #8e8e93;
  --text-secondary-dark: #8e8e93;
  --shadow-subtle: rgba(0, 0, 0, 0.04);
  --shadow-medium: rgba(0, 0, 0, 0.08);
  --shadow-strong: rgba(0, 0, 0, 0.15);
}

/* === Chat Container === */
.chat-container {
  display: flex;
  flex-direction: column;
  width: 100%;
  height: calc(100vh - 16px);
  max-height: 700px;
  background: var(--tg-theme-bg-color, var(--chat-bg-light));
  border-radius: 12px;
  overflow: hidden;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.02), 0 1px 2px rgba(0, 0, 0, 0.01);
  margin: 0 auto;
}

/* Telegram Web (ПК) чат */
body.tg-light-theme .chat-container {
  background: var(--tg-theme-bg-color, var(--chat-bg-light));
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.02), 0 1px 2px rgba(0, 0, 0, 0.01);
}

body.tg-dark-theme .chat-container {
  background: var(--tg-theme-bg-color, var(--chat-bg-dark));
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.1), 0 1px 3px rgba(0, 0, 0, 0.06);
}

@media (prefers-color-scheme: dark) {
  .chat-container {
    background: var(--tg-theme-bg-color, var(--chat-bg-dark));
    box-shadow: 0 4px 24px rgba(0, 0, 0, 0.2);
  }
}

/* === Chat Header === */
.chat-header {
  padding: 12px 16px;
  background: var(--tg-theme-button-color, var(--message-my-bg));
  color: var(--tg-theme-button-text-color, #ffffff);
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-shrink: 0;
  box-shadow: 0 1px 4px rgba(0, 122, 255, 0.08);
}

.chat-header-title {
  font-weight: 600;
  font-size: 18px;
  letter-spacing: -0.3px;
}
.chat-header-info {
  font-size: 14px;
  opacity: 0.9;
  font-weight: 500;
}
.end-chat-button {
  background: rgba(255, 255, 255, 0.15);
  border: none;
  color: white;
  padding: 8px 12px;
  border-radius: 8px;
  cursor: pointer;
  font-size: 13px;
  font-weight: 500;
  transition: all 0.3s ease;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.08);
}

.end-chat-button:hover {
  background: rgba(255, 255, 255, 0.25);
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.12);
}

/* === Messages Area === */
.messages-area {
  flex: 1;
  overflow-y: auto;
  padding: 16px;
  display: flex;
  flex-direction: column;
  gap: 12px;
  background: var(--messages-bg-light);
}

body.tg-light-theme .messages-area {
  background: var(--tg-theme-bg-color, var(--messages-bg-light));
}

body.tg-dark-theme .messages-area {
  background: var(--messages-bg-dark);
}

.messages-area-placeholder {
  text-align: center;
  color: var(--tg-theme-text-color, var(--text-secondary-light));
  padding: 20px 16px;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.4;
}

body.tg-dark-theme .messages-area-placeholder {
  color: var(--tg-theme-text-color, var(--text-secondary-dark));
}

@media (prefers-color-scheme: dark) {
  .messages-area-placeholder {
    color: var(--tg-theme-text-color, var(--text-secondary-dark));
  }
}

/* === Message Bubble === */
.message-row {
  display: flex;
  align-items: flex-end;
  gap: 8px;
}
.message-row.from-me {
  flex-direction: row-reverse;
}
.message-bubble {
  max-width: 80%;
  padding: 10px 14px;
  border-radius: 16px;
  word-break: break-word;
  font-size: 15px;
  line-height: 1.4;
  position: relative;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.03), 0 1px 2px rgba(0, 0, 0, 0.02);
}

.message-bubble.from-me {
  background: var(--tg-theme-button-color, var(--message-my-bg));
  color: var(--tg-theme-button-text-color, #ffffff);
  border-bottom-right-radius: 6px;
  box-shadow: 0 1px 4px rgba(0, 122, 255, 0.1), 0 1px 2px rgba(0, 0, 0, 0.04);
}

.message-bubble.from-partner {
  background: var(--message-partner-bg-light);
  color: var(--tg-theme-text-color, var(--text-primary-light));
  border-bottom-left-radius: 6px;
}

body.tg-light-theme .message-bubble.from-partner {
  background: var(--message-partner-bg-light);
  color: var(--tg-theme-text-color, var(--text-primary-light));
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.03), 0 1px 2px rgba(0, 0, 0, 0.02);
}

body.tg-dark-theme .message-bubble.from-partner {
  background: var(--message-partner-bg-dark);
  color: var(--tg-theme-text-color, var(--text-primary-dark));
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.1), 0 1px 3px rgba(0, 0, 0, 0.06);
}

.message-time {
  font-size: 12px;
  opacity: 0.6;
  margin-top: 6px;
  text-align: right;
  color: inherit;
  font-weight: 400;
}

/* === Typing Indicator === */
.typing-indicator {
  display: flex;
  align-items: center;
  gap: 12px;
  color: var(--tg-theme-text-color, var(--text-secondary-light));
  font-size: 14px;
  padding: 0 8px;
}

body.tg-dark-theme .typing-indicator {
  color: var(--tg-theme-text-color, var(--text-secondary-dark));
}

.typing-indicator-bubble {
  padding: 10px 16px;
  background: var(--message-partner-bg-light);
  border-radius: 18px;
  box-shadow: 0 2px 12px var(--shadow-subtle);
}

body.tg-light-theme .typing-indicator-bubble {
  background: var(--message-partner-bg-light);
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.05);
}

body.tg-dark-theme .typing-indicator-bubble {
  background: var(--message-partner-bg-dark);
  box-shadow: 0 2px 12px rgba(0, 0, 0, 0.2);
}

/* === Input Area === */
.input-area {
  display: flex;
  align-items: center;
  padding: 20px 24px;
  gap: 16px;
  background: var(--input-area-bg-light);
  box-shadow: 0 -2px 16px var(--shadow-subtle);
}

body.tg-light-theme .input-area {
  background: var(--tg-theme-bg-color, var(--input-area-bg-light));
  box-shadow: 0 -1px 8px rgba(0, 0, 0, 0.04);
}

body.tg-dark-theme .input-area {
  background: var(--input-area-bg-dark);
  box-shadow: 0 -2px 16px rgba(0, 0, 0, 0.15);
}

.message-input {
  flex: 1;
  border: none;
  border-radius: 20px;
  padding: 14px 20px;
  font-size: 15px;
  background: var(--input-bg-light);
  color: var(--tg-theme-text-color, var(--text-primary-light));
  transition: all 0.3s ease;
  box-shadow: 0 2px 12px var(--shadow-subtle);
}

body.tg-light-theme .message-input {
  background: var(--input-bg-light);
  color: var(--tg-theme-text-color, var(--text-primary-light));
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.04);
}

body.tg-dark-theme .message-input {
  background: var(--input-bg-dark);
  color: var(--tg-theme-text-color, var(--text-primary-dark));
  box-shadow: 0 2px 12px rgba(0, 0, 0, 0.15);
}

.message-input:focus {
  outline: none;
  background: var(--tg-theme-bg-color, #ffffff);
  box-shadow: 0 4px 20px rgba(0, 122, 255, 0.15);
  transform: translateY(-1px);
}

body.tg-dark-theme .message-input:focus {
  background: #48484a;
  box-shadow: 0 4px 20px rgba(0, 122, 255, 0.25);
}

.message-input::placeholder {
  color: var(--tg-theme-hint-color, var(--text-secondary-light));
}

body.tg-dark-theme .message-input::placeholder {
  color: var(--tg-theme-hint-color, var(--text-secondary-dark));
}

.send-button {
  width: 44px;
  height: 44px;
  border-radius: 50%;
  border: none;
  background: var(--tg-theme-button-color, var(--message-my-bg));
  color: var(--tg-theme-button-text-color, #ffffff);
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  flex-shrink: 0;
  font-size: 18px;
  transition: all 0.3s ease;
  box-shadow: 0 3px 16px rgba(0, 122, 255, 0.2);
}

.send-button:hover {
  transform: translateY(-1px);
  box-shadow: 0 6px 24px rgba(0, 122, 255, 0.3);
}

.send-button:disabled {
  background: #8e8e93;
  cursor: not-allowed;
  opacity: 0.6;
  transform: none;
  box-shadow: 0 2px 8px rgba(142, 142, 147, 0.2);
}

.send-button svg {
  width: 20px;
  height: 20px;
} 