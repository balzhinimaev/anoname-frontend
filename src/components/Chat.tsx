import React, { useState, useEffect, useRef } from 'react';
import websocketService from '../services/websocket';
import { hapticFeedback } from '../utils/telegram';
import {
  WebSocketChatMessageReceived,
  WebSocketChatTyping,
  WebSocketUser
} from '../types';

interface ChatProps {
  chatId: string;
  partnerInfo: {
    telegramId: string;
    gender: 'male' | 'female';
    age: number;
  };
  currentUser: WebSocketUser | null;
  onEndChat: () => void;
}

interface ChatMessage {
  id: string;
  content: string;
  timestamp: string;
  isFromMe: boolean;
  sender: WebSocketUser;
}

const Chat: React.FC<ChatProps> = ({ chatId, partnerInfo, currentUser, onEndChat }) => {
  const [messages, setMessages] = useState<ChatMessage[]>([]);
  const [messageInput, setMessageInput] = useState<string>('');
  const [isTyping, setIsTyping] = useState<boolean>(false);
  const [partnerTyping, setPartnerTyping] = useState<boolean>(false);
  const messagesEndRef = useRef<HTMLDivElement>(null);
  const typingTimeoutRef = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => {
    // –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –æ–±—Ä–∞–±–æ—Ç—á–∏–∫–æ–≤ WebSocket
    websocketService.onChatMessage((data: WebSocketChatMessageReceived) => {
      if (data.chatId === chatId) {
        const newMessage: ChatMessage = {
          id: data.message._id,
          content: data.message.content,
          timestamp: data.message.timestamp,
          isFromMe: currentUser ? data.message.sender.telegramId === currentUser.telegramId : false,
          sender: data.message.sender
        };
        
        setMessages(prev => [...prev, newMessage]);
        
        // –¢–∞–∫—Ç–∏–ª—å–Ω–∞—è –æ–±—Ä–∞—Ç–Ω–∞—è —Å–≤—è–∑—å –¥–ª—è –≤—Ö–æ–¥—è—â–∏—Ö —Å–æ–æ–±—â–µ–Ω–∏–π
        if (!newMessage.isFromMe) {
          hapticFeedback('light');
        }
      }
    });

    websocketService.onChatTyping((data: WebSocketChatTyping) => {
      if (data.chatId === chatId && currentUser && data.userId !== currentUser.telegramId.toString()) {
        setPartnerTyping(true);
        
        // –£–±–∏—Ä–∞–µ–º –∏–Ω–¥–∏–∫–∞—Ç–æ—Ä —á–µ—Ä–µ–∑ 3 —Å–µ–∫—É–Ω–¥—ã
        setTimeout(() => setPartnerTyping(false), 3000);
      }
    });

    // –û—á–∏—Å—Ç–∫–∞ –ø—Ä–∏ —Ä–∞–∑–º–æ–Ω—Ç–∏—Ä–æ–≤–∞–Ω–∏–∏
    return () => {
      if (typingTimeoutRef.current) {
        clearTimeout(typingTimeoutRef.current);
      }
    };
  }, [chatId, currentUser]);

  useEffect(() => {
    // –ü—Ä–æ–∫—Ä—É—Ç–∫–∞ –∫ –ø–æ—Å–ª–µ–¥–Ω–µ–º—É —Å–æ–æ–±—â–µ–Ω–∏—é
    messagesEndRef.current?.scrollIntoView({ behavior: 'smooth' });
  }, [messages]);

  const handleSendMessage = (): void => {
    if (!messageInput.trim()) return;

    try {
      websocketService.sendMessage(chatId, messageInput.trim());
      setMessageInput('');
      setIsTyping(false);
      // –°–±—Ä–∞—Å—ã–≤–∞–µ–º —Ç–∞–π–º–∞—É—Ç, —á—Ç–æ–±—ã –∏–Ω–¥–∏–∫–∞—Ç–æ—Ä "–ø–µ—á–∞—Ç–∞–µ—Ç" –Ω–µ –æ—Ç–ø—Ä–∞–≤–∏–ª—Å—è –ø–æ—Å–ª–µ —Å–æ–æ–±—â–µ–Ω–∏—è
      if (typingTimeoutRef.current) {
        clearTimeout(typingTimeoutRef.current);
      }
      hapticFeedback('light');
    } catch (error) {
      console.error('–û—à–∏–±–∫–∞ –æ—Ç–ø—Ä–∞–≤–∫–∏ —Å–æ–æ–±—â–µ–Ω–∏—è:', error);
      hapticFeedback('error');
    }
  };

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const value = e.target.value;
    setMessageInput(value);

    // –û—Ç–ø—Ä–∞–≤–ª—è–µ–º —Å–æ–±—ã—Ç–∏–µ –Ω–∞–±–æ—Ä–∞ —Ç–µ–∫—Å—Ç–∞
    if (value.length > 0 && !isTyping) {
      setIsTyping(true);
      websocketService.sendTyping(chatId);
    }

    // –°–±—Ä–∞—Å—ã–≤–∞–µ–º —Ñ–ª–∞–≥ –Ω–∞–±–æ—Ä–∞ —á–µ—Ä–µ–∑ 2 —Å–µ–∫—É–Ω–¥—ã –±–µ–∑–¥–µ–π—Å—Ç–≤–∏—è
    if (typingTimeoutRef.current) {
      clearTimeout(typingTimeoutRef.current);
    }

    typingTimeoutRef.current = setTimeout(() => {
      setIsTyping(false);
    }, 2000);
  };

  const handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>): void => {
    if (e.key === 'Enter') {
      e.preventDefault();
      handleSendMessage();
    }
  };

  const handleEndChat = (): void => {
    try {
      websocketService.endChat(chatId, 'user_ended');
      onEndChat();
      hapticFeedback('warning');
    } catch (error) {
      console.error('–û—à–∏–±–∫–∞ –∑–∞–≤–µ—Ä—à–µ–Ω–∏—è —á–∞—Ç–∞:', error);
    }
  };

  const formatTime = (timestamp: string): string => {
    const date = new Date(timestamp);
    return date.toLocaleTimeString('ru-RU', { 
      hour: '2-digit', 
      minute: '2-digit' 
    });
  };

  const SendIcon = () => (
    <svg viewBox="0 0 24 24" fill="currentColor">
      <path d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z" />
    </svg>
  );

  return (
    <div className="chat-container">
      {/* –ó–∞–≥–æ–ª–æ–≤–æ–∫ —á–∞—Ç–∞ */}
      <div className="chat-header">
        <div className="chat-header-info">
          {partnerInfo.gender === 'male' ? 'üë® –ú—É–∂—Å–∫–æ–π' : 'üë© –ñ–µ–Ω—Å–∫–∏–π'}, {partnerInfo.age} –ª–µ—Ç
        </div>
        <button 
          onClick={handleEndChat}
          className="end-chat-button"
        >
          –ó–∞–≤–µ—Ä—à–∏—Ç—å
        </button>
      </div>

      {/* –û–±–ª–∞—Å—Ç—å —Å–æ–æ–±—â–µ–Ω–∏–π */}
      <div className="messages-area">
        {messages.length === 0 && (
          <div className="messages-area-placeholder">
            –û—Ç–ø—Ä–∞–≤—å—Ç–µ –ø–µ—Ä–≤–æ–µ —Å–æ–æ–±—â–µ–Ω–∏–µ üëã
          </div>
        )}

        {messages.map((message) => (
          <div
            key={message.id}
            className={`message-row ${message.isFromMe ? 'from-me' : ''}`}
          >
            <div
              className={`message-bubble ${message.isFromMe ? 'from-me' : 'from-partner'}`}
            >
              <div>{message.content}</div>
              <div className="message-time">
                {formatTime(message.timestamp)}
              </div>
            </div>
          </div>
        ))}

        {partnerTyping && (
          <div className="typing-indicator">
            <div className="typing-indicator-bubble">
              –°–æ–±–µ—Å–µ–¥–Ω–∏–∫ –ø–µ—á–∞—Ç–∞–µ—Ç...
            </div>
          </div>
        )}
        <div ref={messagesEndRef} />
      </div>

      {/* –ü–æ–ª–µ –≤–≤–æ–¥–∞ */}
      <div className="input-area">
        <input
          type="text"
          value={messageInput}
          onChange={handleInputChange}
          onKeyPress={handleKeyPress}
          placeholder="–í–≤–µ–¥–∏—Ç–µ —Å–æ–æ–±—â–µ–Ω–∏–µ..."
          className="message-input"
        />
        <button
          onClick={handleSendMessage}
          disabled={!messageInput.trim()}
          className="send-button"
        >
          <SendIcon />
        </button>
      </div>
    </div>
  );
};

export default Chat; 