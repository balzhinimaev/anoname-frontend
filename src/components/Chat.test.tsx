// @ts-nocheck
import React from 'react';
import renderer from 'react-test-renderer';
import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import Chat from './Chat';

// Мокаем утилиты, чтобы избежать ошибок в среде Vitest
vi.mock('../utils/telegram', () => ({
  hapticFeedback: vi.fn(),
}));

// Мокаем WebSocket сервис
vi.mock('../services/websocket', () => ({
  default: {
    onChatMessage: vi.fn(),
    onChatStartTyping: vi.fn(),
    onChatStopTyping: vi.fn(),
    onChatEnd: vi.fn(),
    sendChatMessage: vi.fn(),
    sendTypingStart: vi.fn(),
    sendTypingStop: vi.fn(),
    sendChatEnd: vi.fn(),
  },
}));

// Мокаем хранилища
vi.mock('../store/chatStore', () => ({
  useChatStore: () => ({
    currentChatId: null,
    partner: null,
    messages: [],
    messageInput: '',
    isTyping: false,
    updateMessageInput: vi.fn(),
    sendMessage: vi.fn(),
    endChat: vi.fn(),
  }),
}));

describe('components/Chat', () => {
  it('renders correctly', () => {
    const { container } = render(<Chat />);
    expect(container).toMatchSnapshot();
  });
}); 