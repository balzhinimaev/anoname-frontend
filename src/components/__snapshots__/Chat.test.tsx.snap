// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/Chat > renders correctly 1`] = `
<div>
  <div
    class="chat-container"
  >
    <div
      class="chat-header"
    >
      Загрузка чата...
    </div>
  </div>
</div>
`;

exports[`components/Chat renders correctly 1`] = `
<div
  className="chat-container"
>
  <div
    className="chat-header"
  >
    Загрузка чата...
  </div>
</div>
`;
